File: tests/python/test_ui.py
"""
Python binding tests for UI (Ulcer Index) indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close
from rust_comparison import compare_with_rust


class TestUI:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_ui_partial_params(self, test_data):
        """Test UI with partial parameters - mirrors check_ui_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.ui(close, period=14, scalar=100.0)
        assert len(result) == len(close)
    
    def test_ui_accuracy(self, test_data):
        """Test UI matches expected values from Rust tests - mirrors check_ui_accuracy"""
        close = test_data['close']
        
        result = ta_indicators.ui(close, period=14, scalar=100.0)
        
        assert len(result) == len(close)
        
        
        expected_last_five = [
            3.514342861283708,
            3.304986039846459,
            3.2011859814326304,
            3.1308860017483373,
            2.909612553474519,
        ]
        
        
        assert_close(
            result[-5:], 
            expected_last_five,
            rtol=1e-6,
            msg="UI last 5 values mismatch"
        )
        
        
        period = 14
        for i in range(period - 1):
            assert np.isnan(result[i]), f"Expected NaN at index {i} during warmup period"
        
        
        params = {'period': 14, 'scalar': 100.0}
        compare_with_rust('ui', result, 'close', params)
    
    def test_ui_default_candles(self, test_data):
        """Test UI with default parameters - mirrors check_ui_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.ui(close, period=14, scalar=100.0)
        assert len(result) == len(close)
    
    def test_ui_zero_period(self):
        """Test UI fails with zero period - mirrors check_ui_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ui(input_data, period=0, scalar=100.0)
    
    def test_ui_period_exceeds_length(self):
        """Test UI fails when period exceeds data length - mirrors check_ui_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ui(data_small, period=10, scalar=100.0)
    
    def test_ui_very_small_dataset(self):
        """Test UI fails with insufficient data - mirrors check_ui_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ui(single_point, period=14, scalar=100.0)
    
    def test_ui_empty_input(self):
        """Test UI fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty|Empty"):
            ta_indicators.ui(empty, period=14, scalar=100.0)
    
    def test_ui_all_nan_input(self):
        """Test UI fails with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.ui(all_nan, period=14, scalar=100.0)
    
    def test_ui_different_scalars(self, test_data):
        """Test UI with different scalar values"""
        close = test_data['close']
        
        
        result1 = ta_indicators.ui(close, period=14, scalar=50.0)
        assert len(result1) == len(close)
        
        
        result2 = ta_indicators.ui(close, period=14, scalar=200.0)
        assert len(result2) == len(close)
        
        
        
        
        result_default = ta_indicators.ui(close, period=14, scalar=100.0)
        for i in range(14*2-2, len(close)):  
            if not np.isnan(result_default[i]) and not np.isnan(result2[i]):
                ratio = result2[i] / result_default[i]
                assert abs(ratio - 2.0) < 0.01, f"Scalar scaling incorrect at index {i}"
    
    def test_ui_kernel_parameter(self, test_data):
        """Test UI with different kernel parameters"""
        close = test_data['close']
        
        
        result1 = ta_indicators.ui(close, period=14, scalar=100.0, kernel=None)
        assert len(result1) == len(close)
        
        
        result2 = ta_indicators.ui(close, period=14, scalar=100.0, kernel="scalar")
        assert len(result2) == len(close)
        
        
        assert_close(result1, result2, rtol=1e-10, msg="Kernel results mismatch")
    
    def test_ui_batch_operations(self, test_data):
        """Test UI batch operations"""
        close = test_data['close']
        
        
        batch_result = ta_indicators.ui_batch(
            close,
            period_range=(14, 14, 1),
            scalar_range=(100.0, 100.0, 0.0)
        )
        
        
        single_result = ta_indicators.ui(close, period=14, scalar=100.0)
        
        assert 'values' in batch_result
        assert 'periods' in batch_result
        assert 'scalars' in batch_result
        
        batch_values = batch_result['values']
        assert batch_values.shape == (1, len(close))
        
        
        assert_close(
            batch_values[0], 
            single_result, 
            rtol=1e-10, 
            msg="Batch vs single mismatch"
        )
    
    def test_ui_batch_multiple_periods(self, test_data):
        """Test UI batch with multiple period values"""
        close = test_data['close'][:100]  
        
        
        batch_result = ta_indicators.ui_batch(
            close,
            period_range=(10, 14, 2),
            scalar_range=(100.0, 100.0, 0.0)
        )
        
        assert batch_result['values'].shape == (3, 100)
        assert len(batch_result['periods']) == 3
        assert list(batch_result['periods']) == [10, 12, 14]
        
        
        for i, period in enumerate([10, 12, 14]):
            single_result = ta_indicators.ui(close, period=period, scalar=100.0)
            assert_close(
                batch_result['values'][i], 
                single_result, 
                rtol=1e-10, 
                msg=f"Period {period} mismatch"
            )
    
    def test_ui_batch_multiple_scalars(self, test_data):
        """Test UI batch with multiple scalar values"""
        close = test_data['close'][:100]  
        
        
        batch_result = ta_indicators.ui_batch(
            close,
            period_range=(14, 14, 1),
            scalar_range=(50.0, 150.0, 50.0)
        )
        
        assert batch_result['values'].shape == (3, 100)
        assert len(batch_result['scalars']) == 3
        assert_close(
            batch_result['scalars'], 
            [50.0, 100.0, 150.0], 
            rtol=1e-10
        )
    
    def test_ui_streaming(self):
        """Test UI streaming functionality"""
        
        stream = ta_indicators.UiStream(period=5, scalar=100.0)
        
        
        
        data = [10.0, 12.0, 8.0, 15.0, 11.0, 9.0, 13.0, 10.0, 14.0, 12.0, 11.0, 13.0, 15.0, 14.0, 16.0]
        
        results = []
        for value in data:
            result = stream.update(value)
            results.append(result)
        
        
        non_none_count = sum(1 for r in results if r is not None)
        assert non_none_count > 0, f"Stream should produce some non-None values. Results: {results}"
        
        
        for i, result in enumerate(results):
            if result is not None:
                assert result >= 0, f"UI at index {i} should be non-negative, got {result}"
                assert result < 1000, f"UI at index {i} should be reasonable, got {result}"
    
    def test_ui_nan_handling(self, test_data):
        """Test UI handles NaN values correctly"""
        close = test_data['close'].copy()
        
        
        close[50:55] = np.nan
        
        
        result = ta_indicators.ui(close, period=14, scalar=100.0)
        assert len(result) == len(close)
        
        
        valid_count = np.sum(~np.isnan(result[60:]))
        assert valid_count > 0, "Should have valid values after NaN region"


if __name__ == "__main__":
    pytest.main([__file__])

File: tests/wasm/test_ui.js
/**
 * WASM binding tests for UI (Ulcer Index) indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('UI partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ui_js(close, 14, 100.0);
    assert.strictEqual(result.length, close.length);
});

test('UI accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ui_js(close, 14, 100.0);
    
    assert.strictEqual(result.length, close.length);
    
    
    const expectedLastFive = [
        3.514342861283708,
        3.304986039846459,
        3.2011859814326304,
        3.1308860017483373,
        2.909612553474519,
    ];
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLastFive,
        1e-6,
        "UI last 5 values mismatch"
    );
    
    
    const warmupPeriod = 14 * 2 - 2; 
    for (let i = 0; i < warmupPeriod; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} during warmup period`);
    }
    
    
    await compareWithRust('ui', result, 'close', { period: 14, scalar: 100.0 });
});

test('UI default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ui_js(close, 14, 100.0);
    assert.strictEqual(result.length, close.length);
});

test('UI zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ui_js(inputData, 0, 100.0);
    }, /Invalid period/);
});

test('UI period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ui_js(dataSmall, 10, 100.0);
    }, /Invalid period/);
});

test('UI very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.ui_js(singlePoint, 14, 100.0);
    }, /Invalid period/);
});

test('UI empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ui_js(empty, 14, 100.0);
    }, /Empty|empty/);
});

test('UI all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.ui_js(allNaN, 14, 100.0);
    }, /All values are NaN/);
});

test('UI different scalars', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const result1 = wasm.ui_js(close, 14, 50.0);
    assert.strictEqual(result1.length, close.length);
    
    
    const result2 = wasm.ui_js(close, 14, 200.0);
    assert.strictEqual(result2.length, close.length);
    
    
    
    const resultDefault = wasm.ui_js(close, 14, 100.0);
    for (let i = 14 * 2 - 2; i < close.length; i++) { 
        if (!isNaN(resultDefault[i]) && !isNaN(result2[i])) {
            const ratio = result2[i] / resultDefault[i];
            assert(Math.abs(ratio - 2.0) < 0.01, `Scalar scaling incorrect at index ${i}`);
        }
    }
});

test('UI batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    const batchResult = wasm.ui_batch(close, {
        period_range: [14, 14, 0],
        scalar_range: [100.0, 100.0, 0]
    });
    
    
    const singleResult = wasm.ui_js(close, 14, 100.0);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
});

test('UI batch multiple periods', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.ui_batch(close, {
        period_range: [10, 14, 2],
        scalar_range: [100.0, 100.0, 0]
    });
    
    
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    
    
    const periods = [10, 12, 14];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);
        
        const singleResult = wasm.ui_js(close, periods[i], 100.0);
        assertArrayClose(
            rowData, 
            singleResult, 
            1e-10, 
            `Period ${periods[i]} mismatch`
        );
    }
});

test('UI batch multiple scalars', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.ui_batch(close, {
        period_range: [14, 14, 0],
        scalar_range: [50.0, 150.0, 50.0]
    });
    
    
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.combos.length, 3);
    
    
    assert.strictEqual(batchResult.combos[0].scalar, 50.0);
    assert.strictEqual(batchResult.combos[1].scalar, 100.0);
    assert.strictEqual(batchResult.combos[2].scalar, 150.0);
});

test('UI batch metadata from result', () => {
    
    
    const close = new Float64Array(30);
    close.fill(100);
    
    const result = wasm.ui_batch(close, {
        period_range: [10, 14, 2],     
        scalar_range: [50.0, 100.0, 50.0] 
    });
    
    
    assert.strictEqual(result.combos.length, 6);
    
    
    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[0].scalar, 50.0);
    
    
    assert.strictEqual(result.combos[5].period, 14);
    assert.strictEqual(result.combos[5].scalar, 100.0);
});


test('UI zero-copy API (in-place)', () => {
    
    const data = new Float64Array([100, 105, 95, 98, 102, 99, 101, 104, 103, 107]);
    const period = 5;
    const scalar = 100.0;
    
    
    const ptr = wasm.ui_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.ui_into(ptr, ptr, data.length, period, scalar);
        
        
        const regularResult = wasm.ui_js(data, period, scalar);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.ui_free(ptr, data.length);
    }
});

test('UI zero-copy with large dataset', () => {
    const size = 100000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = 100 + Math.sin(i * 0.01) * 10 + Math.random() * 2;
    }
    
    const ptr = wasm.ui_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');
    
    try {
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        memView.set(data);
        
        wasm.ui_into(ptr, ptr, size, 14, 100.0);
        
        
        const memView2 = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        
        
        const warmup = 14 * 2 - 2;
        for (let i = 0; i < warmup; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }
        
        
        for (let i = warmup; i < Math.min(warmup + 10, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.ui_free(ptr, size);
    }
});


test('UI zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.ui_into(0, 0, 10, 14, 100.0);
    }, /null pointer|Null pointer/i);
    
    
    const ptr = wasm.ui_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.ui_into(ptr, ptr, 10, 0, 100.0);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.ui_into(ptr, ptr, 10, 15, 100.0);
        }, /Invalid period/);
    } finally {
        wasm.ui_free(ptr, 10);
    }
});


test('UI zero-copy memory management', () => {
    
    const sizes = [100, 1000, 10000, 100000];
    
    for (const size of sizes) {
        const ptr = wasm.ui_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.ui_free(ptr, size);
    }
});


test('UI batch fast API', () => {
    const size = 100;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = 100 + Math.sin(i * 0.1) * 5;
    }
    
    
    const inPtr = wasm.ui_alloc(size);
    assert(inPtr !== 0, 'Failed to allocate input buffer');
    
    
    const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, size);
    inView.set(data);
    
    
    
    const outSize = 6 * size;
    const outPtr = wasm.ui_alloc(outSize);
    assert(outPtr !== 0, 'Failed to allocate output buffer');
    
    try {
        const numCombos = wasm.ui_batch_into(
            inPtr, outPtr, size,
            10, 14, 2,      
            50.0, 100.0, 50.0  
        );
        
        assert.strictEqual(numCombos, 6, 'Expected 6 combinations');
        
        
        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, outSize);
        
        
        let hasNonNaN = false;
        for (let i = 0; i < outSize; i++) {
            if (!isNaN(outView[i])) {
                hasNonNaN = true;
                break;
            }
        }
        assert(hasNonNaN, 'Output should contain some non-NaN values');
    } finally {
        wasm.ui_free(inPtr, size);
        wasm.ui_free(outPtr, outSize);
    }
});

test.after(() => {
    console.log('UI WASM tests completed');
});